### PR TITLE
Add adapter to s3prl frontend

### PR DESCRIPTION
**Pull request by karthik19967829**
This PR adds adapters to wav2vec2 model by adapting s3prl.py 

An updated common voice recipe PR will be created with a sample config

### CODE REVIEW

File: espnet2/asr/frontend/adapter_utils/add_adapters.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    if adapt_layers == []:
+        print(">> adapt_layers is an empy list. No adapters will be inserted.")
+        return
+    orig_param_num = count_params(wav2vec2_model)
+
+    # freeze all layers
+    for param in wav2vec2_model.parameters():
+        param.requires_grad = False
+
+    adapted_layers = []
```
Comment (simpleoier) [inline espnet2/asr/frontend/adapter_utils/add_adapters.py:23]: What is the difference between `adapted _layers` and `adapt_layers`

File: espnet2/asr/frontend/s3prl.py [hunk 2]
```diff
@@ -82,12 +87,26 @@ def _get_upstream(self, frontend_conf):
         ]:
             s3prl_upstream.model.encoder.layerdrop = 0.0
 
+            # check if adapter is added
+            if (
+                s3prl_upstream.model.__class__.__name__ == "Wav2Vec2Model"
+                and self.args.add_adapters
+            ):
+                s3prl_upstream = add_adapters_wav2vec2(
+                    s3prl_upstream, adapter_down_dim=192, adapt_layers=[0, 1, 2]
```
Comment (simpleoier) [inline espnet2/asr/frontend/s3prl.py:96]: I suggest to make the adapter_down_dim, adapt_layers to be configurable.

**Review by simpleoier — COMMENTED** [review]
It looks good and it might work. We need to think about how to make it generalize to other upstreams?

File: espnet2/asr/frontend/s3prl.py (hunk 2 above)
Comment (karthik19967829) [reply espnet2/asr/frontend/s3prl.py:96]: This a very good suggestion , will incorporate it soon

File: espnet2/asr/frontend/adapter_utils/add_adapters.py (hunk 1 above)
Comment (karthik19967829) [reply espnet2/asr/frontend/adapter_utils/add_adapters.py:23]: yes , it seems redundant will get rid of it , and will remove print , I think its better to write into train.log

File: espnet2/asr/frontend/s3prl.py (hunk 2 above)
Comment (karthik19967829) [reply espnet2/asr/frontend/s3prl.py]: Both` adapter_down_dim` and `adapt_layers` are no configurable from the config.yaml file

Comment (karthik19967829) [reply espnet2/asr/frontend/s3prl.py]: @jwrh please test this by , pulling my master into your branch and run it till training starts to ensure the pipeline works. Also try changing the `adapter_down_dim' and `adapt_layers` to a different check if the model created is accurate

Comment (karthik19967829) [reply espnet2/asr/frontend/s3prl.py]: @simpleoier we plan to add support to HUBERT and wavLM in the next PR. 
This PR we would like to keep to wav2vec2 and configurable adapter input.

File: doc/conf.py [hunk 3]
```diff
@@ -81,9 +81,9 @@ def setup(app):
 master_doc = "index"
```
Comment (sw005320) [inline doc/conf.py:81]: please do not change unrelated part.

File: egs/libri_css/asr1/diarization/VB_diarization.py [hunk 4]
```diff
@@ -130,11 +130,11 @@ def VB_diarization(
```
Comment (sw005320) [inline egs/libri_css/asr1/diarization/VB_diarization.py:130]: ditto

File: egs2/commonvoice/asr1/README.md [hunk 5]
```diff
@@ -1,5 +1,105 @@
...
+
+
+
+
+
+
+
+
+
 ## How to use fusion of front-ends :
```
Comment (sw005320) [inline egs2/commonvoice/asr1/README.md:103]:
```suggestion
# How to use fusion of front-ends : 
```
This part is not related to the adapter. So, change the section level to make it independent.

File: egs2/commonvoice/asr1/README.md [hunk 6]
```diff
@@ -1,5 +1,105 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (sw005320) [inline egs2/commonvoice/asr1/README.md:101]: Remove unnecessary line breaks.

**Review by sw005320 — COMMENTED** [review]
I have some minor comments.

- [ ] please add a unit test or integration test given the answer from @simpleoier.

@simpleoier, could you check the details of the network descriptions and their behaviors? Also, what is the best way of adding a test for this kind of frontend implementation? Can you give us some suggestions?

File: espnet2/asr/frontend/s3prl.py [hunk 7]
```diff
@@ -82,12 +87,26 @@ def _get_upstream(self, frontend_conf):
         ]:
             s3prl_upstream.model.encoder.layerdrop = 0.0
 
+            # check if adapter is added
```
Comment (karthik19967829) [inline espnet2/asr/frontend/s3prl.py:90]: @sw005320 @simpleoier this section takes care of first checking if the upstream selected is wav2vec2 before checking for adapter config and adding it . This ensures it does not break when other upstreams are used.

File: espnet2/asr/frontend/s3prl.py [hunk 8]
```diff
@@ -82,12 +87,26 @@ def _get_upstream(self, frontend_conf):
...
+                )
+
         from s3prl.upstream.interfaces import Featurizer
 
         if self.multilayer_feature:
             feature_selection = "hidden_states"
         else:
             feature_selection = "last_hidden_state"
+
+        # check if adapters are added and over-ride the feature selection
```
Comment (karthik19967829) [inline espnet2/asr/frontend/s3prl.py:106]: @simpleoier to avoid dimension issues , when adapter is enabled ,` feature_selection ` is force set to "last_hidden_state"

File: doc/conf.py (hunk 3 above)
Comment (karthik19967829) [reply doc/conf.py:81]: black formatting issue , it changed few unrelated files , will remove it soon.

File: egs/libri_css/asr1/diarization/VB_diarization.py (hunk 4 above)
Comment (karthik19967829) [reply egs/libri_css/asr1/diarization/VB_diarization.py:130]: black formatting issue , it changed few unrelated files , will remove it soon.

File: egs2/commonvoice/asr1/README.md (hunk 5 above)
Comment (karthik19967829) [reply egs2/commonvoice/asr1/README.md:103]: @jwrh

File: egs2/commonvoice/asr1/README.md (hunk 6 above)
Comment (karthik19967829) [reply egs2/commonvoice/asr1/README.md:101]: @jwrh

**Review by sw005320 — COMMENTED** [review]
- I put some comments about the README.md files.
- Please add the unit test (check https://github.com/espnet/espnet/tree/master/test/espnet2/asr/frontend)

@simpleoier, can I ask you to check the main source part again?
Is it the right place to put the adapter-related codes under `frontend/adapter_utils/adapters`? Maybe `frontend/adapters` would be fine.

File: espnet2/asr/frontend/adapter_utils/adapters/adapter.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        if not is_fairseq_available:
+            raise ImportError(
+                "`fairseq` is not available. Please install it via `pip install"
+                " fairseq` or `cd /path/to/espnet/tools && . ./activate_python.sh"
+                " && ./installers/install_fairseq.sh`."
+            )
+
+        self.down_projection = nn.Linear(orig_dim, down_dim)
+        self.up_projection = nn.Linear(down_dim, orig_dim)
+        nn.init.xavier_uniform_(self.down_projection.weight)
```
Comment (simpleoier) [inline espnet2/asr/frontend/adapter_utils/adapters/adapter.py:38]: Does this initialization really work as expected? I feel this [init](https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tasks/asr.py#L517) will overwrite them.

File: espnet2/asr/frontend/s3prl.py (hunk 7 above)
Comment (simpleoier) [reply espnet2/asr/frontend/s3prl.py:90]: I suggest to have a unified interface for "add_adapter". Then in that interface (a function or a class), you can have apply corresponding adapters.

**Review by simpleoier — COMMENTED** [review]
I made a few comments @karthik19967829.

File: egs2/commonvoice/asr1/README.md [hunk 10]
```diff
@@ -1,6 +1,106 @@
 <!-- Generated by scripts/utils/show_asr_result.sh -->
 
-## How to use fusion of front-ends : 
+
+# Results for adapter experiments with commonvoice French dataset and commonvoice Welsh dataset:
```
[the remark this answers is not in the document]
Comment (karthik19967829) [reply egs2/commonvoice/asr1/README.md:4]: Screen Shot 2022-08-12 at 6 27 41 AM
For French we have matched the numbers reported by the Oxford ICASSP 2022 paper  https://arxiv.org/pdf/2202.03218.pdf

File: espnet2/asr/frontend/adapter_utils/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+### description
```
[the remark this answers is not in the document]
Comment (karthik19967829) [reply espnet2/asr/frontend/adapter_utils/README.md:1]: sure will move it , actually we don't need these details for end-user , will add documentation how to use adapters from config.yaml

File: espnet2/asr/frontend/adapter_utils/adapters/adapter.py (hunk 9 above)
Comment (karthik19967829) [reply espnet2/asr/frontend/adapter_utils/adapters/adapter.py:38]: good catch , could you explain while  [https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tasks/asr.py#L517](init) is fired when init is not , None ?
also as this change needs to be done in asr.py while will impact other users , what's your suggestion ?

Comment (jwrh) [reply espnet2/asr/frontend/adapter_utils/adapters/adapter.py:38]: Hi, I've done a side-by-side comparison regarding the actual use of that two lines of code and the results suggest that the initialization somehow does work? My code tracing leaves me to this line below

https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tasks/asr.py#L518

and subsequent inquiry into the `initialize` function suggests that at the very end, after all initializations are done, there is this **reload all parameters** operation

https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/torch_utils/initialize.py#L99

and that method is implemented by loading from a previous saved `self.pretrained_params`

https://github.com/espnet/espnet/blob/b008ac7d58e9ced1a9f8c89cc85ee69d9e9461ab/espnet2/asr/frontend/s3prl.py#L48

Thus my inquiry leaves me with the conclusion that if we build model without that two lines of initialization, the model would proceed with default (therefore uninitialized adapter modules parameters), pertained s3prl frontend. With that two lines of initialization code, we would first construct and load the entire pretrained s3prl frontend and save to `self.pretrained_params`, thus successfully initializing the parameters.

Comment (karthik19967829) [reply espnet2/asr/frontend/adapter_utils/adapters/adapter.py:38]: @simpleoier can we resolve this or do you have any suggestions to make this part of code more clear?

**Comment by karthik19967829** [comment]
@simpleoier added a unified add_adapter function so that it will be a common interface for all upstreams

File: espnet2/asr/frontend/s3prl.py (hunk 7 above)
Comment (karthik19967829) [reply espnet2/asr/frontend/s3prl.py:90]: added a unified adapter function , within the function its checks for upstream type and then applied the corresponding model operation

File: espnet2/asr/frontend/adapter_utils/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (karthik19967829) [reply espnet2/asr/frontend/adapter_utils/README.md]: removed this place , will update espnet docs on using adapters in the next PR

File: espnet2/asr/frontend/adapter_utils/adapters/adapter.py (hunk 9 above)
Comment (simpleoier) [reply espnet2/asr/frontend/adapter_utils/adapters/adapter.py:38]: Yes, it is resolved. I think @jwrh 's tracing explains all my concerns.
